packages/repository/src/relations/has-many/has-many-through-repository.factory.ts

```diff
@@ -0,0 +1,219 @@
+ * between models. For example, if a Customer model is related to an Order model
+ * via a HasMany relation, then, the relational repository returned by the
+ * factory function would be constrained by a Customer model instance's id(s).
+ *
+ * @param relationMetadata The relation metadata used to describe the
+ * relationship and determine how to apply the constraint.
+ * @param targetRepositoryGetter The repository which represents the target model of a
+ * relation attached to a datasource.
+ * @returns The factory function which accepts a foreign key value to constrain
+ * the given target repository
+ */
+export function createHasManyThroughRepositoryFactory<
+  Target extends Entity,
+  TargetID,
+  Through extends Entity,
+  ThroughID,
+  ForeignKeyType
+>(
+  relationMetadata: HasManyThroughDefinition,
+  targetRepositoryGetter: Getter<EntityCrudRepository<Target, TargetID>>,
+  throughRepositoryGetter: Getter<EntityCrudRepository<Through, ThroughID>>,
+): HasManyThroughRepositoryFactory<Target, Through, ForeignKeyType> {
+  const meta = resolveHasManyThroughMetadata(relationMetadata);
+  debug('Resolved HasMany relation metadata: %o', meta);
+  return function(fkValue?: ForeignKeyType) {
+    async function getAdvancedConstraint(
+      targetInstance: Target,
+    ): Promise<AdvancedConstraint<Target | Through>> {
+      return createAdvancedConstraint<
+        Target,
+        Through,
+        ThroughID,
+        ForeignKeyType
+      >(targetInstance, meta, throughRepositoryGetter, fkValue);
+    }
+    return new DefaultHasManyThroughRepository<
+      Target,
+      TargetID,
+      EntityCrudRepository<Target, TargetID>,
+      Through,
+      ThroughID,
+      EntityCrudRepository<Through, ThroughID>
+    >(targetRepositoryGetter, throughRepositoryGetter, getAdvancedConstraint);
+  };
+}
+
+type HasManyThroughResolvedDefinition = HasManyThroughDefinition & {
+  keyTo: string;
+  targetFkName: string;
+  targetPrimaryKey: string;
+};
+
+async function createAdvancedConstraint<
+  Target extends Entity,
+  Through extends Entity,
+  ThroughID,
+  ForeignKeyType
+>(
+  targetInstance: Target,
+  meta: HasManyThroughResolvedDefinition,
+  throughRepositoryGetter: Getter<EntityCrudRepository<Through, ThroughID>>,
+  fkValue?: ForeignKeyType,
+): Promise<AdvancedConstraint<Target | Through>> {
+  // tslint:disable-next-line:no-any
+  const advancedConstraint: any = {
+    dataObject: {[meta.keyTo]: fkValue} as DataObject<Through>,
+    filter: {},
+    where: {},
+  };
+  if (targetInstance) {
+    // through constraint
+    advancedConstraint.dataObject[meta.targetFkName] =
+      targetInstance[meta.targetPrimaryKey as keyof Target];
+    advancedConstraint.where = {where: advancedConstraint.dataObject};
+  } else {
+    // target constraint
+    const throughRepo = await throughRepositoryGetter();
+    const throughInstances = await throughRepo.find(
+      constrainFilter(undefined, advancedConstraint.dataObject),
```

Comment:
I am concerned about this design, where we are making SQL queries as part of the step creating the constraint.

In the future, I would like to implement relational queries using SQL JOINs to get better performance. To be able to do that in a backwards-compatible way, we need to keep all database calls inside the relation repository classes.

Can we come up with such design that `HasManyThroughRepository` accepts a set of constraints (one for the "through" model and another for the "target" model) and all SQL queries are executed by the repository only?

I understand that to build the "target" constraint, we need the data returned by the query of the "through" table. The challenge is to come up with a declarative format for the "target" constraint that will tell the HasManyThroughRepository how to build "target" constraint from "through" query results.

For example, we may decide to compose the constraint from the following fields:
 - a `where` condition for the "through" table (e.g. `{customerId: 1}`)
 - an object mapping "through" columns to "target" columns (e.g. `{sellerId: 'id'}`)
 - an optional `where` condition for the "target" table with additional constraints (e.g. `{country: 'us'}` to constraint the repository to US sellers only)

The HasManyThroughRepository can works as follows:

```ts
const throughInstances = await throughRepo.find(...);
const targetWheres = throughInstances.map(inst => {
  const where = {};
  for (const throughProp in mappingProvidedByConstraint) {
    const targetProp = mappingProvidedByConstraint[throughProp];
    where[targetProp] = inst[throughProp];
  }
  return where;
});
const targetFilter = {where: {
  and: [
    {or: targetWheres}, 
    {...whereProvidedByConstraint},
  ],
  // (I am not sure if I got this filter syntax right)
}};
const result = await targetRepo.find(targetFilter);
```

Another benefit of this approach is that we can get rid of constraint getter and pass the directly the constraint object, the same way as we do for other relation types.

Reply:
Interesting idea. I will need to look into this more.

Reply:
The resolution of hasManyThrough relation for `find` requires the following input:

1. A filter to constrain the source model instances
2. A PK/FK constraint to connect source instances to through instances
3. A PK/FK constraint to connect through instances to target instances
4. A filter to constrain the target model instances

The implementation can be a waterfall of multiple queries or a single join if such models are backed by the same relational database.

Reply:
> The implementation can be a waterfall of multiple queries or a single join if such models are backed by the same relational database.

+1

What's important for me: we need to design our API to allow the implementation to choose whether the operation is implemented via multiple queries or a single JOIN. The currently proposed solution does not offer the flexibility needed, because a query is executed while building one of the constraints.

I'd like us to find a solution how to express constraints declaratively and defer any database queries to the actual repository operation. My comment above offers one of possible solutions.

Reply:
Take a look at my following comment. Perhaps it is best to come back to this later.

https://github.com/strongloop/loopback-next/pull/2359#issuecomment-475979639

Reply:
I suggest adding the PK/FK constraints in a separate pull request. I created a feature request for it here #3545.